v3_async_wip/v3_async_wip/sastoken.py

```diff
@@ -5,173 +5,222 @@
+        :param signing_mechanism: The signing mechanism that will be used to sign data
+        :type signing mechanism: :class:`SigningMechanism`
+        :param int ttl: Time to live for generated tokens, in seconds (default 3600)
         """
-        self._uri = uri
-        self._signing_mechanism = signing_mechanism
-        self._key_name = key_name
-        # These two values will be set by the .refresh() call
-        self._expiry_time: int
-        self._token: str
-
+        self.signing_mechanism = signing_mechanism
+        self.uri = uri
         self.ttl = ttl
-        self.refresh()
-
-    def __str__(self) -> str:
-        return self._token
 
-    def refresh(self) -> None:
-        """
-        Refresh the SasToken lifespan, giving it a new expiry time, and generating a new token.
-        """
-        self._expiry_time = int(time.time() + self.ttl)
-        self._token = self._build_token()
+    async def generate_sastoken(self) -> SasToken:
+        """Generate a new SasToken
 
-    def _build_token(self) -> str:
-        """Build SasToken representation
-
-        :returns: String representation of the token
+        :raises: SasTokenError if the token cannot be generated
         """
-        url_encoded_uri = urllib.parse.quote(self._uri, safe="")
-        message = url_encoded_uri + "\n" + str(self.expiry_time)
+        expiry_time = int(time.time()) + self.ttl
+        url_encoded_uri = urllib.parse.quote(self.uri, safe="")
+        message = url_encoded_uri + "\n" + str(expiry_time)
         try:
-            signature = self._signing_mechanism.sign(message)
+            signature = await self.signing_mechanism.sign(message)
         except Exception as e:
             # Because of variant signing mechanisms, we don't know what error might be raised.
             # So we catch all of them.
-            raise SasTokenError("Unable to build SasToken from given values") from e
+            raise SasTokenError("Unable to generate SasToken") from e
         url_encoded_signature = urllib.parse.quote(signature, safe="")
-        if self._key_name:
-            token = self._auth_rule_token_format.format(
-                resource=url_encoded_uri,
-                signature=url_encoded_signature,
-                expiry=str(self.expiry_time),
-                keyname=self._key_name,
-            )
-        else:
-            token = self._simple_token_format.format(
-                resource=url_encoded_uri,
-                signature=url_encoded_signature,
-                expiry=str(self.expiry_time),
-            )
-        return token
+        token_str = TOKEN_FORMAT.format(
+            resource=url_encoded_uri,
+            signature=url_encoded_signature,
+            expiry=str(expiry_time),
+        )
+        return SasToken(token_str)
 
-    @property
-    def expiry_time(self) -> int:
-        """Expiry Time is READ ONLY"""
-        return self._expiry_time
 
+class ExternalSasTokenGenerator(SasTokenGenerator):
```

Comment:
I'm not thrilled with this "inheriting" from SasTokenGenerator, and then overriding everything. It's mostly for typing - ideally SasTokenGenerator and ExternalSasTokenGenerator would both inherit from an abstract class, which ideally would be called SasTokenGenerator for good typehint semantics, but then what does that leave from us to call the basic child generator class that uses a signing mechanism? SigningMechanismSasTokenGenerator? That's bad.

Open to ideas.

Reply:
I've seen the convention where SasTokenGenerator is the interface and SasTokenGeneratorImpl is the implementation of that interface. We don't have to use this everywhere but maybe here it makes sense.

Reply:
It wasn't officially added until 3.8, but if it's mostly for Typing, then maybe it should be a Protocol instead of an Interface.